cleanup/cleanup.sh

```diff
@@ -1,3 +1,94 @@
-#!/bin/sh
+#!/bin/bash
 
-echo "Plugin called."
+set -e -o pipefail -u
+
+GIT_REPO_PATH=${KUBECTL_PLUGINS_LOCAL_FLAG_GIT_REPO_PATH}
+FORCE=${KUBECTL_PLUGINS_LOCAL_FLAG_FORCE}
+KEEP=${KUBECTL_PLUGINS_LOCAL_FLAG_KEEP}
+FROM_HEAD=$(( KEEP+1 ))
+NAMESPACE=${KUBECTL_PLUGINS_CURRENT_NAMESPACE}
+oc="${KUBECTL_PLUGINS_CALLER} -n ${NAMESPACE}"
+
+get_commit_list() {
+    # all commits except HEAD in the Git repository
+    git -C ${GIT_REPO_PATH} rev-list --author-date-order HEAD | tail -n "+2"
+}
+
+get_active_images() {
+    local dc_rc_sfs_path='{range .items[*]}{.spec.template.spec.containers[*].image}{"\n"}{end}'
+    local pod_cont_path='{range .items[*]}{.spec.containers[*].image}{"\n"}{end}'
+    local pod_init_path='{range .items[*]}{.spec.initContainers[*].image}{"\n"}{end}'
+    local cron_path='{range .items[*]}{.spec.jobTemplate.spec.template.spec.containers[*].image}{"\n"}{end}'
+
+    local images_in_dc_rc_sset=$(${oc} get dc,rc,statefulset -o jsonpath="${dc_rc_sfs_path}")
+    local images_in_pod_cont=$(${oc} get pod -o jsonpath="${pod_cont_path}")
+    local images_in_pod_init=$(${oc} get pod -o jsonpath="${pod_init_path}")
+    local images_in_cron=$(${oc} get cronjob -o jsonpath="${cron_path}")
+
+    echo "${images_in_dc_rc_sset}" \
+         "${images_in_pod_cont}" \
+         "${images_in_pod_init}" \
+         "${images_in_cron}" | sort --unique
+}
+
+get_inactive_images() {
+    local istags_path='{range .items[*]}{.metadata.name}{"\n"}{end}'
+    local sort_by_date='--sort-by .metadata.creationTimestamp'
+
+    local all_tags=$(${oc} get istag -o jsonpath="${istags_path}" ${sort_by_date} \
+                   | tail -n "+${FROM_HEAD}" | sort --unique)
+
+    local all_active=$(get_active_images)
+
+    # all_inactive := all_tags - all_active
+    comm -23 <(echo ${all_tags}) <(echo ${all_active})
+}
+
+get_deletion_candidates() {
+    local commit_list=($(get_commit_list))
+    local inactive_images=($(get_inactive_images))
+    local deletion_candidates=()
+
+    for commit_hash in ${commit_list[*]}; do
+        for candidate in ${inactive_images[*]}; do
+            # remove all images ending with commit SHA
+            if [[ ${candidate} == *${commit_hash} ]]; then
```

Comment:
Here could be a place to implement the filter logic: Instead of `*${commit_hash}`, compare with `${image_name}:${commit_hash}`.

Reply:
After a second thought, my approach would break the `keep` functionality how it is implemented now.